clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1469,20 +1471,58 @@ event_delay_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t
                                  opnd_create_reg(scratch)));
     }
     MINSERT(bb, instr, INSTR_CREATE_jcc(drcontext, OP_jl, opnd_create_instr(skip_call)));
+#        elif defined(AARCH64)
+    if (!drx_insert_counter_update(drcontext, bb, instr,
+                                   (dr_spill_slot_t)(SPILL_SLOT_MAX + 1) /*use drmgr*/,
+                                   (dr_spill_slot_t)(SPILL_SLOT_MAX + 1), &instr_count,
+                                   num_instrs, DRX_COUNTER_64BIT))
+        DR_ASSERT(false);
+
+    reg_id_t scratch1, scratch2, scratch3;
+    if (drreg_reserve_register(drcontext, bb, instr, NULL, &scratch1) != DRREG_SUCCESS ||
+        drreg_reserve_register(drcontext, bb, instr, NULL, &scratch2) != DRREG_SUCCESS ||
+        drreg_reserve_register(drcontext, bb, instr, NULL, &scratch3) != DRREG_SUCCESS)
+        FATAL("Fatal error: failed to reserve scratch registers");
+    dr_save_arith_flags_to_reg(drcontext, bb, instr, scratch3);
```

Comment:
From an initial exploration, I found that the problem with drreg_reserve_aflags() is that at some point (for a long run where `-trace_after_instrs` is set to a very large value) the aflags are not restored. By forcing aflags to restore in drreg_insert_restore_all()  (essentially by replacing the if statement in https://github.com/DynamoRIO/dynamorio/blob/f5016906b5699773acf52dd7f5147da89f10ae12/ext/drreg/drreg.c#L411 with  just `if (!pt->aflags.native)`) , the issue is resolved and I can use drreg_reserve_aflags(). Now the question is at what point the aflags should have been restored and didn't, and which of the conditions in lines 412-418 should have been true and weren't.

Reply:
What is the problem exactly? Do you get a crash? From my experience, this issue often arises more due to a bug exposed by the instructions in a particular basic block, rather than the length of the trace. For example, drreg mistakenly expects a given instruction to overwrite eflags, and therefore restoration is not done. 

Such problems can be a bit of a pain to debug. If you do get a crash, you can ask DR to output logs (with log level 3 or above), and inspect the few basic blocks right before the crash to maybe find anything suspicious.

Reply:
true the error is irrelevant to the length of the trace. The bug exhibited in a big workload after million of instructions probably because of some specific sequence of instructions that is not that common. I have not been able to trigger a similar issue in smaller benchmarks or any test.

The problem is that the program takes at some point some unexpected control flow and is eventually led to an application-level assertion down the line. I looked at the logs but they are too long and there is nothing suspicious at the last few basic blocks. Probably the clobbered aflags affect something much earlier than the end. I cannot pinpoint where things went wrong, but I might be able to find out where the aflags were not restored and look into those cases. That said, so far I have been more successful looking at the drreg code and trying to reason about its correctness rather than investigating the problematic run.

Reply:
Comparing the block tags between a bad and good run, w/ ASLR or other nondet sources turned off, is one method we have used to find divergence points, but that still requires walking backward from the differing conditional branch to the block where the bad flags were produced: but for flags this should be quite close, as opposed to GPR's.

Here, adding flags checks to ensure the app flags value is the same before the instru and at the next sink point (app flags read or end of block) should help.  Xref #4678 on library support to make such things easier.

Reply:
Got it. I will try comparing the block tags to figure out the issue.

Reply:
It took some time but this approach worked. There were various points of divergence without any noticeable error related to flags. Divergence was caused by content differences in general-purpose registers. Thankfully, I noticed several basic blocks before a point of divergence unrecognized instructions (`xx`) that executed before the flags restoration. If any of these instructions read/wrote the flags then the application would behave incorrectly. I filed an issue. See #4687 for more info. With the #4688 fix, the issue is resolved for AArch64.

Reply:
With the #4688 fix merged, I was able to use `drreg_reserve_aflags()`.